File: src/api/getStudyGroups.ts
import axios from 'axios';
import { KeyValue } from '.';

export const getStudyGroups = async (facultyId: string, educationFormId: string, courseId: string): Promise<KeyValue[]> => {
    
    const response = await axios(`https://vnz.osvita.net/WidgetSchedule.asmx/GetStudyGroups`, {
        params: {
            callback: '',
            aVuzID: 11613,
            aFacultyID: `"${facultyId}"`,
            aEducationForm: `"${educationFormId}"`,
            aCourse: `"${courseId}"`,     
            aGiveStudyTimes: false,
        },
    });
    const data = response.data.d?.studyGroups;
    if (!data) throw new Error('Failed to get study groups');
    
    return data;

}

File: api/get_study_groups.go
package api

import (
	"encoding/json"
	"io"
	"net/http"
)

type getStudyGroupsResponse struct {
	D struct {
		StudyGroups []KeyValue `json:"studyGroups"`
	} `json:"d"`
}

func GetStudyGroups(faculty string, educationForm string, course string) ([]KeyValue, error) {
	req, _ := http.NewRequest("GET", "https://vnz.osvita.net/WidgetSchedule.asmx/GetStudyGroups", nil)

	q := req.URL.Query()
	q.Add("callback", "")
	q.Add("aVuzId", "11613")
	q.Add("aGiveStudyTimes", "false")
	q.Add("aFacultyID", "\""+faculty+"\"")
	q.Add("aEducationForm", "\""+educationForm+"\"")
	q.Add("aCourse", "\""+course+"\"")
	req.URL.RawQuery = q.Encode()

	res, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}

	body, err := io.ReadAll(res.Body)
	if err != nil {
		return nil, err
	}

	var result getStudyGroupsResponse
	err = json.Unmarshal(body, &result)
	if err != nil {
		return nil, err
	}

	SortKeyValue(result.D.StudyGroups)

	return result.D.StudyGroups, nil
}
